zilliqa/src/contracts/deposit.sol

```diff
@@ -138,61 +207,238 @@ contract Deposit {
+            revert("A slashed validator cannot self-suspend.");
+        }
+
+        // Note: we immediately set the validator status to suspended here.
+        // This is very generous to the validators: they can simply pause operation at any point, with no penalty.
+        // An alternative would be to keep the validator Active. Then it would be removed from consensus at the
+        // next epoch boundary (a fast-track compared to simply withdrawing all stake, which would take up to 2 epochs),
+        // but would remain liable for liveness until then.
+        _stakersMap[blsPubKey].status = Status.Suspended;
+
+        Value storage newSuspension = kickQueue.pushBack();
+        newSuspension.key = blsPubKey;
+        // no associated amount
+    }
+
+    function eject(
+        bytes calldata blsPubKey,
+        uint256 slashAmount
+    ) public isStaker() {
+        bytes storage callerKey = _stakerKeys[msg.sender];
+        Staker storage caller = _stakersMap[callerKey];
+
+        // Check if the call is authorised.
+        if (caller.stakedBalance < (totalStake / 10)) {
+            revert(
+                "call must come from a reward address corresponding to a staker with more than 10% stake"
+            );
+        }
+
+        // set status to slashed, put into kickqueue for next epoch
+        _stakersMap[blsPubKey].status = Status.Slashed;
+        // If a validator is not currently in the committee and has status == Slashed,
+        // it should not be possible for it to get into the committee in any way.
+        // Thus we only bother queueing a kick if it's currently in the committee.
+        if (_stakersMap[blsPubKey].keyIndex >= 0) {
+            Value storage ejection = kickQueue.pushBack();
+            ejection.key = blsPubKey;
+            ejection.amount = slashAmount;
+        }
+    }
+
+    function swapRemoveValidator(uint index) private {
+        _stakersMap[committeeKeys[index]].keyIndex = -1;
+        if (index != committeeKeys.length - 1) {
+            bytes storage lastValidatorKey = committeeKeys[committeeKeys.length - 1];
+            committeeKeys[index] = lastValidatorKey;
+            _stakersMap[lastValidatorKey].keyIndex = int(index);
+        }
+        committeeKeys.pop();
+    }
+
+
+    function currentEpoch() public view returns (uint64) {
+        return uint64(block.number / blocksPerEpoch); // TODO - take the value from the shard contract rather than storing it locally
+    }
+
+    function currentQueueIndex() private view returns (uint8) {
+        return uint8(currentEpoch() % WAIT_EPOCHS);
+    }
+
+    function tickEpoch() public {
+        // if current epoch has already been processed, exit
+        if (headProcessedEpoch >= currentEpoch()) {
+            return;
+        }
+
+        uint8 queueIndex = currentQueueIndex();
+
+        // Process pending queue for this epoch
+        for (uint i = 0; i < pendingQueues[queueIndex].length(); ++i) {
```

Comment:
Malicious validators could mount a DoS attack by calling `depositTopup()` a lot of times with small values so that `tickEpoch()` would need to consume more than the block limit iterating through the `pendingQueue`